libgadget/treewalk.c

```diff
@@ -623,14 +600,26 @@ treewalk_run(TreeWalk * tw, int * active_set, size_t size)
     ev_begin(tw, active_set, size);
 
     if(tw->preprocess) {
-        size_t i;
+        int i;
         #pragma omp parallel for
         for(i = 0; i < tw->WorkSetSize; i ++) {
-            const int p_i = tw->WorkSet ? tw->WorkSet[i] : (int) i;
+            const int p_i = tw->WorkSet ? tw->WorkSet[i] : i;
             tw->preprocess(p_i, tw);
         }
     }
 
+    /* First run does all the local interactions*/
+    tw->exportonly = 0;
+    tw->FirstNotExported = 0;
+
+    /* The first iteration of this loop does all local interactions
```

Comment:
I am a bit confused about first loop does all local interactions subsequence skips all local interactions. It appears that ev_primary supposed to be doing all local particles, and ev_second supposed to be only working on imported particles?

It seems to be clearer if there is a way to life the export logic from ev_primary to ev_get_remove(perhaps change name to ev_exchange). Then the loop will explicitly follow what you described here in the comments:

ev_primary(....)
while {
 ev_get_remote()
 ev_secondary()
}

And perhaps the export-only flag and FirstNotExported can be simplified to some extent.

Reply:
I like that idea!

Reply:
I wondered about splitting it like "do_toptree_walk" which finds exports and then doing the rest of the walk in a separate function.

Reply:
Do you plan to use the same `do_toptree_walk` for other activities than exporting?
The usual distributed narrative is "export", "work", "reduce" -- in this narrative, whether the export is from toplevel tree walk or something else is a detail.

Reply:
I was toying with the idea that the do_toptree_walk would also specify which of the local toptree nodes would be used for walking. I think though that "export", "work" and "reduce" are the right verbs and I will use them in the writeup!